packaging/tags.py

```diff
@@ -0,0 +1,382 @@
+        else:
+            return "i386"
+    else:
+        return arch
+
+
+def _mac_binary_formats(version, cpu_arch):
+    formats = [cpu_arch]
+    if cpu_arch == "x86_64":
+        if version >= (10, 4):
+            formats.extend(["intel", "fat64", "fat32"])
+        else:
+            return []
+    elif cpu_arch == "i386":
+        if version >= (10, 4):
+            formats.extend(["intel", "fat32", "fat"])
+        else:
+            return []
+    elif cpu_arch == "ppc64":
+        # TODO: Need to care about 32-bit PPC for ppc64 through 10.2?
+        if version > (10, 5) or version < (10, 4):
+            return []
+        else:
+            formats.append("fat64")
+    elif cpu_arch == "ppc":
+        if version <= (10, 6):
+            formats.extend(["fat32", "fat"])
+        else:
+            return []
+
+    formats.append("universal")
+    return formats
+
+
+def _mac_platforms(version=None, arch=None):
+    version_str, _, cpu_arch = platform.mac_ver()
+    if version is None:
+        version = tuple(map(int, version_str.split(".")[:2]))
+    if arch is None:
+        arch = _mac_arch(cpu_arch)
+    platforms = []
+    for minor_version in range(version[1], -1, -1):
+        compat_version = version[0], minor_version
+        binary_formats = _mac_binary_formats(compat_version, arch)
+        for binary_format in binary_formats:
+            platforms.append(
+                "macosx_{major}_{minor}_{binary_format}".format(
+                    major=compat_version[0],
+                    minor=compat_version[1],
+                    binary_format=binary_format,
+                )
+            )
+    return platforms
+
+
+# From PEP 513.
+def _is_manylinux_compatible(name, glibc_version):
+    # Check for presence of _manylinux module.
+    try:
+        import _manylinux
+
+        return bool(getattr(_manylinux, name + "_compatible"))
+    except (ImportError, AttributeError):
+        # Fall through to heuristic check below.
+        pass
+
+    return _have_compatible_glibc(*glibc_version)
+
+
+# From PEP 513.
+def _have_compatible_glibc(major, minimum_minor):
+    import ctypes
+
+    process_namespace = ctypes.CDLL(None)
+    try:
+        gnu_get_libc_version = process_namespace.gnu_get_libc_version
+    except AttributeError:
+        # Symbol doesn't exist -> therefore, we are not linked to
+        # glibc.
+        return False
+
+    # Call gnu_get_libc_version, which returns a string like "2.5".
+    gnu_get_libc_version.restype = ctypes.c_char_p
+    version_str = gnu_get_libc_version()
+    # py2 / py3 compatibility:
+    if not isinstance(version_str, str):
+        version_str = version_str.decode("ascii")
+
+    # Parse string and check against requested version.
+    version = [int(piece) for piece in version_str.split(".")]
+    assert len(version) == 2
+    if major != version[0]:
+        return False
+    if minimum_minor > version[1]:
+        return False
+    return True
```

Comment:
You should re-sync this code with pip's version -- pip got some bug fixes that didn't make it back into the PEP. (In particular, it turns out that there are weird glibc version strings in the wild that will crash this code.)

Reply:
You'll also probably need to provide a public API for just fetching the glibc version string, since pip needs that for the metrics it uploads to pypi, and it doesn't make sense to have two separate copies of the ctypes code floating around.

Reply:
I'll pull pip's code for this over.

As for a public API, I think that can wait until after this PR is merged as that's tangential to what I'm trying to accomplish (or put another way, I don't want to hold up this PR to design it 😄 ).